Index: mysql-5.0.67/include/my_base.h
===================================================================
--- mysql-5.0.67.orig/include/my_base.h
+++ mysql-5.0.67/include/my_base.h
@@ -41,6 +41,15 @@
 #endif	/* stdin */
 #include <my_list.h>
 
+/*
+  Clean workaround for inconsistent BIG_TABLES define issue
+  https://bugs.launchpad.net/ourdelta/+bug/284123
+  http://bugs.mysql.com/35346
+  Author: Antony Curtis, patch created by Arjen Lentz
+*/
+#undef BIG_TABLES
+#define BIG_TABLES 1
+
 /* The following is bits in the flag parameter to ha_open() */
 
 #define HA_OPEN_ABORT_IF_LOCKED		0	/* default */
Index: mysql-5.0.67/patch_info/big_tables_fixlp284123_fixmysql35346.info
===================================================================
--- /dev/null
+++ mysql-5.0.67/patch_info/big_tables_fixlp284123_fixmysql35346.info
@@ -0,0 +1,6 @@
+File=big_tables_fixlp284123_fixmysql35346.patch
+Name=Fix for http://bugs.mysql.com/35346 (BIG_TABLES define)
+Version=1.0
+Author=Antony Curtis <atcurtis@mac.com>
+License=N/A
+Comment=Patch created by Arjen Lentz <arjen@openquery.com>
