--- Percona-Server-5.5.24-rel26.0/scripts/mysqlhotcopy.sh~	2012-04-27 16:02:58.000000000 +0300
+++ Percona-Server-5.5.24-rel26.0/scripts/mysqlhotcopy.sh	2012-07-20 15:04:53.541258700 +0300
@@ -802,8 +802,8 @@
 
     my $tables =
         eval {
-            $dbh->selectall_arrayref('SHOW TABLES FROM ' .
-                                     $dbh->quote_identifier($db))
+            $dbh->selectall_arrayref('SHOW FULL TABLES FROM ' .
+                                     $dbh->quote_identifier($db) . 'WHERE Table_type = "BASE TABLE"')
         } || [];
     warn "Unable to retrieve list of tables in $db: $@" if $@;
 
