--- Percona-Server-5.5.33-rel31.1/scripts/mysqlhotcopy.sh~	2013-10-02 20:29:57.000000000 +0300
+++ Percona-Server-5.5.33-rel31.1/scripts/mysqlhotcopy.sh	2013-10-02 20:30:33.825747635 +0300
@@ -268,6 +268,7 @@
     $sth_dbs->execute;
     while ( my ($db_name) = $sth_dbs->fetchrow_array ) {
 	next if $db_name =~ m/^information_schema$/i;
+	next if $db_name =~ m/^performance_schema$/i;
 	push @db_desc, { 'src' => $db_name, 't_regex' => $t_regex } if ( $db_name =~ m/$opt{regexp}/o );
     }
 }
