diff -Naur mysql-3.23.25-beta/Docs/Makefile.am mysql-3.23.25-beta-p/Docs/Makefile.am
--- mysql-3.23.25-beta/Docs/Makefile.am	Wed Sep 27 10:51:44 2000
+++ mysql-3.23.25-beta-p/Docs/Makefile.am	Thu Oct 12 21:40:23 2000
@@ -18,7 +18,7 @@
 			Support/generate-mirror-listing.pl
 info_TEXINFOS =		manual.texi
 
-targets =		manual.txt mysql.info manual.html manual.pdf
+targets =		mysql.info manual.html
 
 BUILT_SOURCES =		$(targets) manual_toc.html include.texi
 EXTRA_DIST =		$(noinst_SCRIPTS) $(BUILT_SOURCES) mysqld_error.txt INSTALL-BINARY
--- mysql-3.23.25-beta/Docs/Makefile.in	Wed Sep 27 11:02:20 2000
+++ mysql-3.23.25-beta-p/Docs/Makefile.in	Thu Oct 12 22:03:10 2000
@@ -168,7 +168,7 @@
 
 info_TEXINFOS = manual.texi
 
-targets = manual.txt mysql.info manual.html manual.pdf
+targets = manual.txt mysql.info manual.html
 
 BUILT_SOURCES = $(targets) manual_toc.html include.texi
 EXTRA_DIST = $(noinst_SCRIPTS) $(BUILT_SOURCES) mysqld_error.txt INSTALL-BINARY
