--- mysql-3.23.49/configure.in~	Mon Jan 22 14:37:03 2001
+++ mysql-3.23.49/configure.in	Sat Feb 10 02:45:27 2001
@@ -622,11 +622,7 @@
   AC_CHECK_LIB(nsl_r, gethostbyname_r, [],
     AC_CHECK_LIB(nsl, gethostbyname_r))])
 AC_CHECK_FUNC(setsockopt, , AC_CHECK_LIB(socket, setsockopt))
-AC_CHECK_FUNC(yp_get_default_domain, ,
-  AC_CHECK_LIB(nsl, yp_get_default_domain))
 AC_CHECK_FUNC(p2open, , AC_CHECK_LIB(gen, p2open))
-# This may get things to compile even if bind-8 is installed
-AC_CHECK_FUNC(bind, , AC_CHECK_LIB(bind, bind))
 # For crypt() on Linux
 AC_CHECK_LIB(crypt, crypt)
 AC_CHECK_FUNC(crypt, AC_DEFINE(HAVE_CRYPT)) 
