--- mysql-5.0.90/libmysqld/log.cc~	2010-01-29 14:35:22.000000000 +0200
+++ mysql-5.0.90/libmysqld/log.cc	2010-01-29 14:39:10.268964291 +0200
@@ -2605,11 +2605,11 @@
   bool result=0;
   if (opt_error_log)
   {
+    VOID(pthread_mutex_lock(&LOCK_error_log));
+#ifdef __WIN__
     char err_renamed[FN_REFLEN], *end;
     end= strmake(err_renamed,log_error_file,FN_REFLEN-5);
     strmov(end, "-old");
-    VOID(pthread_mutex_lock(&LOCK_error_log));
-#ifdef __WIN__
     char err_temp[FN_REFLEN+5];
     /*
      On Windows is necessary a temporary file for to rename
@@ -2637,7 +2637,6 @@
      result= 1;
 #else
    FILE *reopen;
-   my_rename(log_error_file,err_renamed,MYF(0));
    if (freopen(log_error_file,"a+",stdout))
        reopen= freopen(log_error_file,"a+",stderr);
    else
--- mysql-5.0.90/sql/log.cc~	2010-01-29 14:41:08.000000000 +0200
+++ mysql-5.0.90/sql/log.cc	2010-01-29 14:42:26.542261420 +0200
@@ -2605,11 +2605,11 @@
   bool result=0;
   if (opt_error_log)
   {
+    VOID(pthread_mutex_lock(&LOCK_error_log));
+#ifdef __WIN__
     char err_renamed[FN_REFLEN], *end;
     end= strmake(err_renamed,log_error_file,FN_REFLEN-5);
     strmov(end, "-old");
-    VOID(pthread_mutex_lock(&LOCK_error_log));
-#ifdef __WIN__
     char err_temp[FN_REFLEN+5];
     /*
      On Windows is necessary a temporary file for to rename
@@ -2635,7 +2635,6 @@
      result= 1;
 #else
    FILE *reopen;
-   my_rename(log_error_file,err_renamed,MYF(0));
    if (freopen(log_error_file,"a+",stdout))
        reopen= freopen(log_error_file,"a+",stderr);
    else

