--- mysql-5.1.6-alpha/storage/myisam/Makefile.am~	2006-02-21 16:08:48.000000000 +0200
+++ mysql-5.1.6-alpha/storage/myisam/Makefile.am	2006-02-21 16:08:51.000000000 +0200
@@ -21,7 +21,7 @@
 LDADD =			@CLIENT_EXTRA_LDFLAGS@ libmyisam.a \
 			$(top_builddir)/mysys/libmysys.a \
 			$(top_builddir)/dbug/libdbug.a \
-			$(top_builddir)/strings/libmystrings.a @ZLIB_LIBS@
+			$(top_builddir)/strings/libmystrings.a @ZLIB_LIBS@ $(libdir)/libsupc++.la
 pkglib_LIBRARIES =	libmyisam.a
 bin_PROGRAMS =		myisamchk myisamlog myisampack myisam_ftdump
 myisamchk_DEPENDENCIES=	$(LIBRARIES)
