--- mysql-5.7.9/sql/CMakeLists.txt~	2015-10-12 08:21:35.000000000 +0200
+++ mysql-5.7.9/sql/CMakeLists.txt	2015-10-22 13:35:37.172783012 +0200
@@ -335,7 +335,7 @@
   )
 ENDIF()
 
-ADD_LIBRARY(sql STATIC ${SQL_SOURCE})
+ADD_CONVENIENCE_LIBRARY(sql STATIC ${SQL_SOURCE})
 ADD_DEPENDENCIES(sql GenServerSource)
 ADD_DEPENDENCIES(sql GenDigestServerSource)
 ADD_DEPENDENCIES(sql GenBootstrapPriv)
@@ -362,15 +362,15 @@
                    log_event.cc log_event_old.cc binlog.cc sql_binlog.cc
                    rpl_filter.cc rpl_record.cc rpl_record_old.cc
                    rpl_utility.cc rpl_injector.cc rpl_table_access.cc)
-ADD_LIBRARY(binlog ${BINLOG_SOURCE})
+ADD_CONVENIENCE_LIBRARY(binlog ${BINLOG_SOURCE})
 TARGET_LINK_LIBRARIES(binlog binlogevents_static)
 
 SET (RPL_SOURCE rpl_handler.cc rpl_tblmap.cc rpl_context.cc)
 ADD_DEPENDENCIES(binlog GenError)
-ADD_LIBRARY(rpl ${RPL_SOURCE})
+ADD_CONVENIENCE_LIBRARY(rpl ${RPL_SOURCE})
 SET (MASTER_SOURCE rpl_master.cc rpl_binlog_sender.cc)
 ADD_DEPENDENCIES(rpl GenError)
-ADD_LIBRARY(master ${MASTER_SOURCE})
+ADD_CONVENIENCE_LIBRARY(master ${MASTER_SOURCE})
 ADD_DEPENDENCIES(master GenError)
 SET (SLAVE_SOURCE rpl_slave.cc rpl_reporting.cc rpl_mi.cc rpl_rli.cc
 		  rpl_info_handler.cc rpl_info_file.cc
@@ -379,7 +379,7 @@
                   rpl_rli_pdb.cc rpl_info_dummy.cc rpl_mts_submode.cc
                   rpl_slave_commit_order_manager.cc rpl_msr.cc
                   rpl_trx_boundary_parser.cc rpl_channel_service_interface.cc)
-ADD_LIBRARY(slave ${SLAVE_SOURCE})
+ADD_CONVENIENCE_LIBRARY(slave ${SLAVE_SOURCE})
 ADD_DEPENDENCIES(slave GenError)
 
 ######################### GUnit Lib #################################
@@ -396,7 +396,7 @@
   LIST(APPEND SQL_GUNIT_LIB_SOURCE named_pipe.cc)
 ENDIF()
 
-ADD_LIBRARY(sqlgunitlib ${SQL_GUNIT_LIB_SOURCE})
+ADD_CONVENIENCE_LIBRARY(sqlgunitlib ${SQL_GUNIT_LIB_SOURCE})
 ADD_DEPENDENCIES(sqlgunitlib GenError)
 
 ######################### GUnit Lib #################################
@@ -558,7 +558,7 @@
 
 
 IF(WIN32 OR HAVE_DLOPEN AND NOT DISABLE_SHARED)
-  ADD_LIBRARY(udf_example MODULE udf_example.cc)
+  ADD_CONVENIENCE_LIBRARY(udf_example MODULE udf_example.cc)
   SET_TARGET_PROPERTIES(udf_example PROPERTIES PREFIX "")
   # udf_example depends on strings 
   IF(WIN32)
