--- Percona-Server-5.6.15-rel63.0/client/mysql.cc.orig	2014-02-09 06:23:16.085128298 +0100
+++ Percona-Server-5.6.15-rel63.0/client/mysql.cc	2014-02-09 06:38:54.561791068 +0100
@@ -103,7 +103,9 @@
 #include <conio.h>
 #else
 #include <readline.h>
+#include <history.h>
 #define HAVE_READLINE
+#define USE_NEW_READLINE_INTERFACE
 #define USE_POPEN
 #endif
   //int vidattr(long unsigned int attrs);	// Was missing in sun curses
@@ -1107,7 +1109,7 @@
 } HIST_ENTRY; 
 #endif
 
-extern "C" int add_history(const char *command); /* From readline directory */
+extern "C" void add_history(const char *command); /* From readline directory */
 extern "C" int read_history(const char *command);
 extern "C" int write_history(const char *command);
 extern "C" HIST_ENTRY *history_get(int num);
