--- mysql/configure.in	2004-02-17 01:20:24.000000000 +0100
+++ mysql/configure.in	2004-02-17 01:21:50.805018856 +0100
@@ -1197,8 +1197,7 @@
 then
   # Look for LinuxThreads.
   AC_MSG_CHECKING("LinuxThreads")
-    res=`grep Linuxthreads /usr/include/pthread.h 2>/dev/null | wc -l`
-    if test "$res" -gt 0
+    if test -f /lib/libpthread.so.0
     then
       AC_MSG_RESULT("Found")
       AC_DEFINE(HAVE_LINUXTHREADS)
