--- mysql-5.1.28-rc/sql/log.cc~	2008-08-28 18:40:55.000000000 +0300
+++ mysql-5.1.28-rc/sql/log.cc	2008-10-27 15:06:13.528944807 +0200
@@ -5021,10 +5021,10 @@
   if (opt_error_log)
   {
     char err_renamed[FN_REFLEN], *end;
-    end= strmake(err_renamed,log_error_file,FN_REFLEN-5);
-    strmov(end, "-old");
     VOID(pthread_mutex_lock(&LOCK_error_log));
 #ifdef __WIN__
+    end= strmake(err_renamed,log_error_file,FN_REFLEN-5);
+    strmov(end, "-old");
     char err_temp[FN_REFLEN+5];
     /*
      On Windows is necessary a temporary file for to rename
@@ -5060,7 +5059,6 @@
     else
      result= 1;
 #else
-   my_rename(log_error_file,err_renamed,MYF(0));
    if (freopen(log_error_file,"a+",stdout))
    {
      FILE *reopen;

