--- Percona-Server-5.6.13-rel61.0/scripts/mysql_config.sh~	2013-10-07 12:25:25.000000000 +0200
+++ Percona-Server-5.6.13-rel61.0/scripts/mysql_config.sh	2013-10-15 18:17:41.547988430 +0200
@@ -124,23 +124,23 @@
 for remove in DDBUG_OFF DSAFE_MUTEX DFORCE_INIT_OF_VARS \
               DEXTRA_DEBUG DHAVE_purify O 'O[0-9]' 'xO[0-9]' 'W[-A-Za-z]*' \
               'mtune=[-A-Za-z0-9]*' 'mcpu=[-A-Za-z0-9]*' 'march=[-A-Za-z0-9]*' \
-              unroll2 ip mp restrict
+              unroll2 ip mp restrict 'f[-a-z0-9]*' 'g[-a-z0-9]*' 'm[-a-z0-9]*'
 do
   # The first option we might strip will always have a space before it because
   # we set -I$pkgincludedir as the first option
-  cflags=`echo "$cflags"|sed -e "s/ -$remove  */ /g"` 
+  cflags=`echo "$cflags"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"`
   cxxflags=`echo "$cxxflags"|sed -e "s/ -$remove  */ /g"` 
 done
 cflags=`echo "$cflags"|sed -e 's/ *\$//'` 
 cxxflags=`echo "$cxxflags"|sed -e 's/ *\$//'` 
 
 # Same for --libs(_r)
-for remove in lmtmalloc static-libcxa i-static static-intel
+for remove in lmtmalloc static-libcxa i-static static-intel 'Wl,-[-a-z0-9,]*'
 do
   # We know the strings starts with a space
-  libs=`echo "$libs"|sed -e "s/ -$remove  */ /g"` 
-  libs_r=`echo "$libs_r"|sed -e "s/ -$remove  */ /g"` 
-  embedded_libs=`echo "$embedded_libs"|sed -e "s/ -$remove  */ /g"` 
+  libs=`echo "$libs"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"`
+  libs_r=`echo "$libs_r"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"`
+  embedded_libs=`echo "$embedded_libs"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"`
 done
 
 # Strip trailing and ending space if any, and '+' (FIXME why?)
