diff -urNp -x '*.orig' mysql-5.0.96.org/Docs/mysql.info mysql-5.0.96/Docs/mysql.info
--- mysql-5.0.96.org/Docs/mysql.info	2012-03-02 15:04:57.000000000 +0100
+++ mysql-5.0.96/Docs/mysql.info	2022-10-18 08:25:51.826064806 +0200
@@ -1,8 +1,9 @@
 This is manual.info, produced by makeinfo version 4.8 from manual.texi.
 
-START-INFO-DIR-ENTRY
-* mysql: (mysql).               MySQL documentation.
-END-INFO-DIR-ENTRY
+@dircategory Data bases:
+@direntry
+* mysql: (mysql).			MySQL documentation
+@end direntry
 
 
 File: manual.info,  Node: Top,  Next: preface,  Prev: (dir),  Up: (dir)
