--- mysql-5.1.49/storage/sphinx/Makefile.am~	2009-02-13 23:26:46.000000000 +0200
+++ mysql-5.1.49/storage/sphinx/Makefile.am	2010-07-24 16:35:57.568716704 +0300
@@ -31,7 +31,7 @@
 LDADD =
 
 DEFS= @DEFS@ \
-      -D_REENTRANT -D_PTHREADS -DENGINE -DSTORAGE_ENGINE -DMYSQL_SERVER
+      -D_REENTRANT -D_PTHREADS -DSTORAGE_ENGINE -DMYSQL_SERVER
 
 noinst_HEADERS =	ha_sphinx.h
 
--- mysql-5.1.49/storage/sphinx/Makefile.am~	2010-07-24 16:37:48.000000000 +0300
+++ mysql-5.1.49/storage/sphinx/Makefile.am	2010-07-24 17:00:56.161565823 +0300
@@ -38,7 +38,7 @@
 EXTRA_LTLIBRARIES =	ha_sphinx.la
 pkgplugin_LTLIBRARIES = @plugin_sphinx_shared_target@ sphinx.la
 
-ha_sphinx_la_LDFLAGS =	-module -rpath $(MYSQLLIBdir)
+ha_sphinx_la_LDFLAGS =	-module -avoid-version -rpath $(MYSQLLIBdir)
 ha_sphinx_la_CXXFLAGS=	$(AM_CFLAGS) -DMYSQL_DYNAMIC_PLUGIN
 ha_sphinx_la_CFLAGS =	$(AM_CFLAGS) -DMYSQL_DYNAMIC_PLUGIN
 ha_sphinx_la_SOURCES =	ha_sphinx.cc
