--- mysql/configure.in	2004-02-17 01:20:24.000000000 +0100
+++ mysql/configure.in	2004-02-17 01:21:50.805018856 +0100
@@ -1273,8 +1273,7 @@
           # old method, check headers
           # Look for LinuxThreads.
           AC_MSG_CHECKING("LinuxThreads in header file comment")
-          res=`grep Linuxthreads /usr/include/pthread.h 2>/dev/null | wc -l`
-          if test "$res" -gt 0
+	  if grep -qs pthread_create /usr/include/pthread.h
           then
             AC_MSG_RESULT("Found")
             AC_DEFINE([HAVE_LINUXTHREADS], [1],

