--- mysql-5.1.26-rc/scripts/mysql_config.sh.orig	2008-07-01 00:36:15.000000000 +0200
+++ mysql-5.1.26-rc/scripts/mysql_config.sh	2008-08-25 19:39:20.111236513 +0200
@@ -132,21 +132,21 @@
               DEXTRA_DEBUG DHAVE_purify O 'O[0-9]' 'xO[0-9]' 'W[-A-Za-z]*' \
               'mtune=[-A-Za-z0-9]*' 'mcpu=[-A-Za-z0-9]*' 'march=[-A-Za-z0-9]*' \
               Xa xstrconst "xc99=none" AC99 \
-              unroll2 ip mp restrict
+              unroll2 ip mp restrict 'f[-a-z0-9]*' 'g[-a-z0-9]*' 'm[-a-z0-9]*'
 do
   # The first option we might strip will always have a space before it because
   # we set -I$pkgincludedir as the first option
-  cflags=`echo "$cflags"|sed -e "s/ -$remove  */ /g"` 
+  cflags=`echo "$cflags"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"` 
 done
 cflags=`echo "$cflags"|sed -e 's/ *\$//'` 
 
 # Same for --libs(_r)
-for remove in lmtmalloc static-libcxa i-static static-intel
+for remove in lmtmalloc static-libcxa i-static static-intel 'Wl,-[-a-z0-9,]*'
 do
   # We know the strings starts with a space
-  libs=`echo "$libs"|sed -e "s/ -$remove  */ /g"` 
-  libs_r=`echo "$libs_r"|sed -e "s/ -$remove  */ /g"` 
-  embedded_libs=`echo "$embedded_libs"|sed -e "s/ -$remove  */ /g"` 
+  libs=`echo "$libs"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"` 
+  libs_r=`echo "$libs_r"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"` 
+  embedded_libs=`echo "$embedded_libs"|sed -e "s/ \(-$remove  *\)\{1,\}/ /g"` 
 done
 
 # Strip trailing and ending space if any, and '+' (FIXME why?)
