diff -urN mysql-4.1.7.org/scripts/mysql_fix_privilege_tables.sh mysql-4.1.7/scripts/mysql_fix_privilege_tables.sh
--- mysql-4.1.7.org/scripts/mysql_fix_privilege_tables.sh	2004-10-27 01:35:56.000000000 +0200
+++ mysql-4.1.7/scripts/mysql_fix_privilege_tables.sh	2004-10-27 12:13:42.159439392 +0200
@@ -5,7 +5,7 @@
 # Default values (Can be changed in my.cnf)
 password=""
 host="localhost"
-user="root"
+user="mysql"
 sql_only=0
 basedir=""
 verbose=0
@@ -15,7 +15,7 @@
 database="mysql"
 bindir=""
 
-file=mysql_fix_privilege_tables.sql
+file=/usr/share/mysql/mysql_fix_privilege_tables.sql
 
 # The following test is to make this script compatible with the 4.0 where
 # the single argument could be a password
