--- mysql-5.1.12-beta/config/ac-macros/ha_ndbcluster.m4~	2006-10-26 15:01:57.000000000 +0300
+++ mysql-5.1.12-beta/config/ac-macros/ha_ndbcluster.m4	2006-11-03 13:40:07.479721306 +0200
@@ -197,7 +197,7 @@
 
   MAKE_BINARY_DISTRIBUTION_OPTIONS="$MAKE_BINARY_DISTRIBUTION_OPTIONS --with-ndbcluster"
 
-  CXXFLAGS="$CXXFLAGS \$(NDB_CXXFLAGS)"
+  CXXFLAGS="$CXXFLAGS $(NDB_CXXFLAGS)"
   if test "$have_ndb_debug" = "default"
   then
     have_ndb_debug=$with_debug
