--- 1.50/scripts/mysqlhotcopy.sh        2005-06-06 19:17:13 -07:00
+++ 1.51/scripts/mysqlhotcopy.sh        2005-11-10 17:05:32 -08:00
@@ -252,6 +252,7 @@
     my $sth_dbs = $dbh->prepare("show databases");
     $sth_dbs->execute;
     while ( my ($db_name) = $sth_dbs->fetchrow_array ) {
+	next if $db_name =~ m/^information_schema$/i;
 	push @db_desc, { 'src' => $db_name, 't_regex' => $t_regex } if ( $db_name =~ m/$opt{regexp}/o );
     }
 }
