--- mysql-3.22.30.pius/find-perl-requires	Thu Jan  1 01:00:00 1970
+++ mysql-3.22.30/find-perl-requires	Fri Feb  4 18:15:29 2000
@@ -0,0 +1,11 @@
+#!/bin/sh
+ulimit -c 0
+
+filelist=`sed "s/['\"]/\\\&/g"`
+
+requires="`echo $filelist|/usr/lib/rpm/find-requires`"
+requires_perl="`/usr/lib/rpm/perl.req $filelist|grep -v 'perl(the)'|grep -v 'perl(mysql)'`"
+requires_mod="`rpm -q --whatprovides --qf "%{NAME}\n"  $requires_perl 2>/dev/null`"
+echo "$requires
+$requires_perl
+$requires_mod"|  sort -u
