--- ./scripts/mysqlhotcopy.sh~	2013-03-19 11:30:02.000000000 +0200
+++ ./scripts/mysqlhotcopy.sh	2013-05-10 00:42:25.986175909 +0300
@@ -789,7 +789,7 @@
 
     eval {
 	my ($file,$position) = get_row( $dbh, "show master status" );
-	die "master status is undefined" if !defined $file || !defined $position;
+	warn "master status is undefined" if !defined $file || !defined $position;
 	
 	my $row_hash = get_row_hash( $dbh, "show slave status" );
 	my ($master_host, $log_file, $log_pos ); 
