diff -urNp -x '*.orig' mysql-5.0.96.org/sql/mysqld.cc mysql-5.0.96/sql/mysqld.cc
--- mysql-5.0.96.org/sql/mysqld.cc	2022-10-18 08:27:46.716064781 +0200
+++ mysql-5.0.96/sql/mysqld.cc	2022-10-18 08:27:47.712731447 +0200
@@ -3766,7 +3766,7 @@ int main(int argc, char **argv)
   select_thread_in_use=1;
 
 #ifdef HAVE_LIBWRAP
-  libwrapName= my_progname+dirname_length(my_progname);
+  libwrapName= (char *)my_progname+dirname_length(my_progname);
   openlog(libwrapName, LOG_PID, LOG_AUTH);
 #endif
 
