--- mysql-5.1.28-rc/sql/log.cc~	2008-08-28 18:40:55.000000000 +0300
+++ mysql-5.1.28-rc/sql/log.cc	2008-10-27 15:06:13.528944807 +0200
@@ -4606,11 +4606,11 @@
   bool result=0;
   if (opt_error_log)
   {
+    VOID(pthread_mutex_lock(&LOCK_error_log));
+#ifdef __WIN__
     char err_renamed[FN_REFLEN], *end;
     end= strmake(err_renamed,log_error_file,FN_REFLEN-4);
     strmov(end, "-old");
-    VOID(pthread_mutex_lock(&LOCK_error_log));
-#ifdef __WIN__
     char err_temp[FN_REFLEN+4];
     /*
      On Windows is necessary a temporary file for to rename
@@ -4642,7 +4642,6 @@
     else
      result= 1;
 #else
-   my_rename(log_error_file,err_renamed,MYF(0));
    if (freopen(log_error_file,"a+",stdout))
      freopen(log_error_file,"a+",stderr);
    else
