--- Percona-Server-5.6.13-rel61.0/sql/CMakeLists.txt~	2013-10-07 12:25:25.000000000 +0200
+++ Percona-Server-5.6.13-rel61.0/sql/CMakeLists.txt	2013-10-15 20:37:28.259765844 +0200
@@ -237,20 +237,20 @@
                    log_event.cc log_event_old.cc binlog.cc sql_binlog.cc
 		   rpl_filter.cc rpl_record.cc rpl_record_old.cc rpl_utility.cc
 		   rpl_injector.cc)
-ADD_LIBRARY(binlog ${BINLOG_SOURCE})
+ADD_LIBRARY(binlog STATIC ${BINLOG_SOURCE})
 SET (RPL_SOURCE rpl_handler.cc rpl_tblmap.cc)
 ADD_DEPENDENCIES(binlog GenError)
-ADD_LIBRARY(rpl ${RPL_SOURCE})
+ADD_LIBRARY(rpl STATIC ${RPL_SOURCE})
 SET (MASTER_SOURCE rpl_master.cc)
 ADD_DEPENDENCIES(rpl GenError)
-ADD_LIBRARY(master ${MASTER_SOURCE})
+ADD_LIBRARY(master STATIC ${MASTER_SOURCE})
 ADD_DEPENDENCIES(master GenError)
 SET (SLAVE_SOURCE rpl_slave.cc rpl_reporting.cc rpl_mi.cc rpl_rli.cc
 		  rpl_info_handler.cc rpl_info_file.cc rpl_info_table.cc
 		  rpl_info_values.cc rpl_info.cc rpl_info_factory.cc
 		  rpl_info_table_access.cc dynamic_ids.cc rpl_rli_pdb.cc
 		  rpl_info_dummy.cc)
-ADD_LIBRARY(slave ${SLAVE_SOURCE})
+ADD_LIBRARY(slave STATIC ${SLAVE_SOURCE})
 ADD_DEPENDENCIES(slave GenError)
 ADD_LIBRARY(sqlgunitlib
   filesort_utils.cc mdl.cc sql_list.cc sql_string.cc thr_malloc.cc
--- percona-server-5.6.28-76.1/storage/tokudb/CMakeLists.txt~	2016-01-27 21:17:38.000000000 +0100
+++ percona-server-5.6.28-76.1/storage/tokudb/CMakeLists.txt	2016-01-28 07:23:41.844001444 +0100
@@ -1,6 +1,7 @@
 SET(TOKUDB_VERSION )
 # PerconaFT only supports x86-64 and cmake-2.8.9+
 IF(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" AND
+    "${CMAKE_SIZEOF_VOID_P}" EQUAL "8" AND
     NOT CMAKE_VERSION VERSION_LESS "2.8.9")
 CHECK_CXX_SOURCE_COMPILES(
 "
--- percona-server-5.6.37-82.2/storage/tokudb/PerconaFT/cmake_modules/TokuSetupCompiler.cmake~	2017-08-24 13:36:46.000000000 +0200
+++ percona-server-5.6.37-82.2/storage/tokudb/PerconaFT/cmake_modules/TokuSetupCompiler.cmake	2017-10-09 20:12:03.039109021 +0200
@@ -197,8 +197,8 @@ if (NOT CMAKE_CXX_COMPILER_ID STREQUAL C
 endif ()
 
 ## always want these
-set(CMAKE_C_FLAGS "-Wall -Werror ${CMAKE_C_FLAGS}")
-set(CMAKE_CXX_FLAGS "-Wall -Werror ${CMAKE_CXX_FLAGS}")
+set(CMAKE_C_FLAGS "-Wall ${CMAKE_C_FLAGS}")
+set(CMAKE_CXX_FLAGS "-Wall ${CMAKE_CXX_FLAGS}")
 
 # pick language dialect
 set(CMAKE_C_FLAGS "-std=c99 ${CMAKE_C_FLAGS}")
--- percona-server-5.6.37-82.2/plugin/tokudb-backup-plugin/Percona-TokuBackup/backup/CMakeLists.txt~	2017-10-09 20:37:05.000000000 +0200
+++ percona-server-5.6.37-82.2/plugin/tokudb-backup-plugin/Percona-TokuBackup/backup/CMakeLists.txt	2017-10-09 20:42:02.664375729 +0200
@@ -18,7 +18,7 @@ set_property(DIRECTORY APPEND PROPERTY C
   _FILE_OFFSET_BITS=64 
   _LARGEFILE64_SOURCE)
 
-set(CMAKE_CXX_FLAGS "-Werror -W -Wall -Wshadow ${CMAKE_CXX_FLAGS}")
+set(CMAKE_CXX_FLAGS "-W -Wall -Wshadow ${CMAKE_CXX_FLAGS}")
 
 set(USE_VALGRIND OFF CACHE BOOL "whether to use valgrind headers")
 if (USE_VALGRIND)

--- percona-server-5.6.51-91.0/include/my_global.h~	2021-02-02 15:20:34.000000000 +0100
+++ percona-server-5.6.51-91.0/include/my_global.h	2024-04-04 10:20:48.289940995 +0200
@@ -380,10 +380,7 @@ C_MODE_END
   @param  member  Name of the member within the structure.
 */
 #define my_container_of(ptr, type, member)              \
-  ({                                                    \
-    const typeof(((type *)0)->member) *__mptr= (ptr);   \
-    (type *)((char *)__mptr - offsetof(type, member));  \
-  })
+      ((type *)((char *)ptr - offsetof(type, member)))

 /*
   A lot of our programs uses asserts, so better to always include it
@@ -519,7 +516,7 @@ typedef int	my_socket;	/* File descripto
 /* Type for fuctions that handles signals */
 #define sig_handler RETSIGTYPE
 C_MODE_START
-typedef void	(*sig_return)();/* Returns type from signal */
+typedef void	(*sig_return)(int);/* Returns type from signal */
 C_MODE_END
 #if defined(__GNUC__) && !defined(_lint)
 typedef char	pchar;		/* Mixed prototypes can take char */
--- percona-server-5.6.51-91.0/configure.cmake~	2021-02-02 15:20:34.000000000 +0100
+++ percona-server-5.6.51-91.0/configure.cmake	2026-05-20 00:00:00.000000000 +0200
@@ -978,7 +978,7 @@ CHECK_C_SOURCE_COMPILES("
 #ifdef __cplusplus
 extern \"C\" void (*signal (int, void (*)(int)))(int);
 #else
-void (*signal ()) ();
+extern void (*signal(int, void (*)(int)))(int);
 #endif
 int main(int ac, char **av) {}
 " SIGNAL_RETURN_TYPE_IS_VOID)
--- percona-server-5.6.51-91.0/storage/tokudb/PerconaFT/portability/toku_assert.h~	2021-02-02 15:20:34.000000000 +0100
+++ percona-server-5.6.51-91.0/storage/tokudb/PerconaFT/portability/toku_assert.h	2026-05-20 00:00:00.000000000 +0200
@@ -43,6 +43,7 @@ Copyright (c) 2006, 2015, Percona and/or
 /* This version will complain if NDEBUG is set. */
 /* It evaluates the argument and then calls a function  toku_do_assert() which takes all the hits for the branches not taken. */

+#undef NDEBUG
 #include <portability/toku_config.h>

 #include <stdint.h>
--- percona-server-5.6.51-91.0/storage/tokudb/PerconaFT/cmake_modules/TokuBuildTagDatabases.cmake~	2021-02-02 15:20:34.000000000 +0100
+++ percona-server-5.6.51-91.0/storage/tokudb/PerconaFT/cmake_modules/TokuBuildTagDatabases.cmake	2026-05-20 00:00:00.000000000 +0200
@@ -37,10 +37,14 @@ option(USE_ETAGS "Build the etags databa
 if (USE_ETAGS)
   find_program(ETAGS "etags")
   if (NOT ETAGS MATCHES NOTFOUND)
+      file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/etags.files" "")
+      foreach(file ${all_srcs} ${all_hdrs})
+          file(APPEND "${CMAKE_CURRENT_BINARY_DIR}/etags.files" "${file}\n")
+      endforeach(file)
     add_custom_command(
       OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/TAGS"
       OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/etags-stamp"
-      COMMAND ${ETAGS} -o TAGS ${all_srcs} ${all_hdrs}
+      COMMAND ${ETAGS} -o TAGS -L "${CMAKE_CURRENT_BINARY_DIR}/etags.files"
       COMMAND touch "${CMAKE_CURRENT_BINARY_DIR}/etags-stamp"
       DEPENDS ${all_srcs} ${all_hdrs} install_tdb_h generate_config_h generate_log_code
       WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}")
@@ -54,10 +58,14 @@ if (USE_CTAGS AND
     (NOT APPLE OR NOT USE_ETAGS))
   find_program(CTAGS "ctags")
   if (NOT CTAGS MATCHES NOTFOUND)
+      file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/ctags.files" "")
+      foreach(file ${all_srcs} ${all_hdrs})
+          file(APPEND "${CMAKE_CURRENT_BINARY_DIR}/ctags.files" "${file}\n")
+      endforeach(file)
     add_custom_command(
       OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/tags"
       OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/ctags-stamp"
-      COMMAND ${CTAGS} -o tags ${all_srcs} ${all_hdrs}
+      COMMAND ${CTAGS} -o tags -L "${CMAKE_CURRENT_BINARY_DIR}/ctags.files"
       COMMAND touch "${CMAKE_CURRENT_BINARY_DIR}/ctags-stamp"
       DEPENDS ${all_srcs} ${all_hdrs} install_tdb_h generate_config_h generate_log_code
       WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}")
