diff -urN mysql-4.0.10-gamma.org/include/my_global.h mysql-4.0.10-gamma/include/my_global.h
--- mysql-4.0.10-gamma.org/include/my_global.h	Wed Jan 29 12:29:19 2003
+++ mysql-4.0.10-gamma/include/my_global.h	Thu Feb 13 14:06:34 2003
@@ -407,7 +407,7 @@
 /* Type for fuctions that handles signals */
 #define sig_handler RETSIGTYPE
 C_MODE_START
-typedef void	(*sig_return)();/* Returns type from signal */
+typedef void	(*sig_return)(int);/* Returns type from signal */
 C_MODE_END
 #if defined(__GNUC__) && !defined(_lint)
 typedef char	pchar;		/* Mixed prototypes can take char */
diff -urN mysql-4.0.10-gamma.org/mysys/mf_iocache.c mysql-4.0.10-gamma/mysys/mf_iocache.c
--- mysql-4.0.10-gamma.org/mysys/mf_iocache.c	Wed Jan 29 12:29:20 2003
+++ mysql-4.0.10-gamma/mysys/mf_iocache.c	Thu Feb 13 13:56:17 2003
@@ -1162,7 +1162,7 @@
   if (info->type == SEQ_READ_APPEND)
   {
     /* Destroy allocated mutex */
-    info->type=0;
+    info->type=(cache_type)0;
 #ifdef THREAD
     pthread_mutex_destroy(&info->append_buffer_lock);
 #endif
diff -urN mysql-4.0.10-gamma.org/mysys/my_thr_init.c mysql-4.0.10-gamma/mysys/my_thr_init.c
--- mysql-4.0.10-gamma.org/mysys/my_thr_init.c	Wed Jan 29 12:29:21 2003
+++ mysql-4.0.10-gamma/mysys/my_thr_init.c	Thu Feb 13 13:53:41 2003
@@ -22,6 +22,8 @@
 #include "mysys_priv.h"
 #include <m_string.h>
 
+#define pthread_mutexattr_setkind_np(x,y)       pthread_mutexattr_settype(x,y)
+
 #ifdef THREAD
 #ifdef USE_TLS
 pthread_key(struct st_my_thread_var*, THR_KEY_mysys);
diff -urN mysql-4.0.10-gamma.org/tools/mysqlmanager.c mysql-4.0.10-gamma/tools/mysqlmanager.c
--- mysql-4.0.10-gamma.org/tools/mysqlmanager.c	Wed Jan 29 12:29:21 2003
+++ mysql-4.0.10-gamma/tools/mysqlmanager.c	Thu Feb 13 13:38:49 2003
@@ -1577,7 +1577,7 @@
 static void manager_exec_free(void* e)
 {
   mysql_close(&((struct manager_exec*)e)->mysql);
-  my_free(e,MYF(0));
+  my_free((char *)e,MYF(0));
 }
 
 static int hex_val(char c)
