diff -urNp -x '*.orig' mysql-5.0.96.org/sql/mysqld.cc mysql-5.0.96/sql/mysqld.cc
--- mysql-5.0.96.org/sql/mysqld.cc	2022-10-18 08:27:46.716064781 +0200
+++ mysql-5.0.96/sql/mysqld.cc	2022-10-18 08:27:47.712731447 +0200
@@ -3766,7 +3766,7 @@ int main(int argc, char **argv)
   select_thread_in_use=1;
 
 #ifdef HAVE_LIBWRAP
-  libwrapName= my_progname+dirname_length(my_progname);
+  libwrapName= (char *)my_progname+dirname_length(my_progname);
   openlog(libwrapName, LOG_PID, LOG_AUTH);
 #endif
 
commit 77fb9608a975ada6ec44d7dfb875967ecdac06a6
Author: Christoffer Hall <hall@mysql.com>
Date:   Mon Jun 22 14:43:19 2009 +0200

    BUG#42721 configure.in invokes 'nm' directly
    
    A hardcoded use of nm and libc was put into configure.in. This broke
    cross-compiling and is bad practice. Fitting variables now used.

diff --git a/configure.in b/configure.in
index bd26e584bfe..33ac021ec76 100644
--- a/configure.in
+++ b/configure.in
@@ -603,10 +603,11 @@ AC_SUBST(NOINST_LDFLAGS)
 # Check if we are using Linux and a glibc compiled with static nss
 # (this is true on the MySQL build machines to avoid NSS problems)
 #
+AC_CHECK_TOOL([NM], [nm]) 
 
 if test "$TARGET_LINUX" = "true" -a "$static_nss" = ""
 then
-  tmp=`nm /usr/lib*/libc.a  | grep _nss_files_getaliasent_r`
+  tmp=`$NM ${other_libc_lib:-/usr/lib*}/libc.a | grep _nss_files_getaliasent_r1`
   if test -n "$tmp"
   then
      STATIC_NSS_FLAGS="-lc -lnss_files -lnss_dns -lresolv"

--- mysql-5.0.96.org/config/ac-macros/alloca.m4	2012-03-02 15:04:09.000000000 +0100
+++ mysql-5.0.96/config/ac-macros/alloca.m4	2022-10-18 10:49:46.076063461 +0200
@@ -45,21 +45,6 @@
    # use ar to extract alloca.o from them instead of compiling alloca.c.
    ALLOCA=alloca.o
    AC_DEFINE(C_ALLOCA, 1)
- 
- AC_CACHE_CHECK(whether alloca needs Cray hooks, ac_cv_os_cray,
- [AC_EGREP_CPP(webecray,
- [#if defined(CRAY) && ! defined(CRAY2)
- webecray
- #else
- wenotbecray
- #endif
- ], ac_cv_os_cray=yes, ac_cv_os_cray=no)])
- if test "$ac_cv_os_cray" = "yes"; then
- for ac_func in _getb67 GETB67 getb67; do
-   AC_CHECK_FUNC($ac_func, [AC_DEFINE_UNQUOTED(CRAY_STACKSEG_END, $ac_func)
-   break])
- done
- fi
  fi
  AC_SUBST(ALLOCA)dnl
 else
