diff -ruN mysql-5.7.21.orig/cmake/readline.cmake mysql-5.7.21/cmake/readline.cmake
--- mysql-5.7.21.orig/cmake/readline.cmake	2017-12-28 04:46:26.000000000 +0100
+++ mysql-5.7.21/cmake/readline.cmake	2018-03-31 21:17:28.373826351 +0200
@@ -110,23 +110,23 @@ ENDMACRO()
 
 MACRO (FIND_SYSTEM_EDITLINE)
   FIND_PATH(FOUND_EDITLINE_READLINE
-    NAMES editline/readline.h
+    NAMES readline/readline.h
   )
   IF(FOUND_EDITLINE_READLINE)
-    SET(EDITLINE_INCLUDE_DIR "${FOUND_EDITLINE_READLINE}/editline")
+    SET(EDITLINE_INCLUDE_DIR "${FOUND_EDITLINE_READLINE}/readline")
   ELSE()
     # Different path on FreeBSD
     FIND_PATH(FOUND_EDIT_READLINE_READLINE
       NAMES edit/readline/readline.h
     )
     IF(FOUND_EDIT_READLINE_READLINE)
-      SET(EDITLINE_INCLUDE_DIR "${FOUND_EDIT_READLINE_READLINE}/edit/readline")
+      SET(EDITLINE_INCLUDE_DIR "${FOUND_EDIT_READLINE_READLINE}/readline")
     ENDIF()
   ENDIF()
 
   FIND_LIBRARY(EDITLINE_LIBRARY
     NAMES
-    edit
+    readline
   )
   MARK_AS_ADVANCED(EDITLINE_INCLUDE_DIR EDITLINE_LIBRARY)
 
@@ -139,7 +139,7 @@ MACRO (FIND_SYSTEM_EDITLINE)
     LIST(APPEND CMAKE_REQUIRED_LIBRARIES ${EDITLINE_LIBRARY})
     CHECK_CXX_SOURCE_COMPILES("
     #include <stdio.h>
-    #include <readline.h>
+    #include <readline/history.h>
     int main(int argc, char **argv)
     {
        HIST_ENTRY entry;
@@ -149,14 +149,13 @@ MACRO (FIND_SYSTEM_EDITLINE)
 
     CHECK_CXX_SOURCE_COMPILES("
     #include <stdio.h>
-    #include <readline.h>
+    #include <readline/readline.h>
     int main(int argc, char **argv)
     {
-      typedef int MYFunction(const char*, int);
-      MYFunction* myf= rl_completion_entry_function;
-      int res= (myf)(NULL, 0);
-      completion_matches(0,0);
-      return res;
+      rl_compentry_func_t *myf= rl_completion_entry_function;
+      char *res= (myf)(NULL, 0);
+      rl_completion_matches(0,0);
+      return res ? 0 : 1;
     }"
     EDITLINE_HAVE_COMPLETION_INT)
 
@@ -168,7 +167,7 @@ MACRO (FIND_SYSTEM_EDITLINE)
       typedef char* MYFunction(const char*, int);
       MYFunction* myf= rl_completion_entry_function;
       char *res= (myf)(NULL, 0);
-      completion_matches(0,0);
+      rl_completion_matches(0,0);
       return res != NULL;
     }"
     EDITLINE_HAVE_COMPLETION_CHAR)
--- mysql-8.1.0/CMakeLists.txt~	2023-06-21 09:52:10.000000000 +0200
+++ mysql-8.1.0/CMakeLists.txt	2023-10-20 10:08:46.673284042 +0200
@@ -20,6 +20,8 @@
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
 
+variable_watch(CMAKE_EXECUTABLE_SUFFIX)
+
 MESSAGE(STATUS "Running cmake version ${CMAKE_VERSION}")
 
 IF(WIN32)
