diff -uNr old-mysql-4.0.14/scripts/mysql_fix_privilege_tables.sh mysql-4.0.14/scripts/mysql_fix_privilege_tables.sh
--- old-mysql-4.0.14/scripts/mysql_fix_privilege_tables.sh	Fri Jul 18 14:57:48 2003
+++ mysql-4.0.14/scripts/mysql_fix_privilege_tables.sh	Wed Oct  1 07:03:28 2003
@@ -2,7 +2,7 @@
 
 root_password="$1"
 host="localhost"
-user="root"
+user="mysql"
 port=""
 socket=""
 comment=""
 database="mysql"
