--- mysql-3.23.37/include/Makefile.am.wiget	Fri May 11 13:23:37 2001
+++ mysql-3.23.37/include/Makefile.am	Fri May 11 13:24:18 2001
@@ -20,12 +20,13 @@
 			mysqld_error.h my_list.h \
 		        my_pthread.h my_no_pthread.h raid.h errmsg.h \
 			my_global.h my_net.h \
+			my_dir.h \
 			sslopt-case.h sslopt-longopts.h sslopt-usage.h \
 			sslopt-vars.h $(BUILT_SOURCES)
 noinst_HEADERS =	global.h config-win.h \
 			nisam.h heap.h merge.h my_bitmap.h\
 			myisam.h myisampack.h myisammrg.h ft_global.h\
-			my_dir.h mysys_err.h my_base.h \
+			mysys_err.h my_base.h \
 			my_nosys.h my_alarm.h queues.h \
 			my_tree.h hash.h thr_alarm.h thr_lock.h \
 			getopt.h t_ctype.h violite.h \
