diff -ur mysql-5.0.19.orig/configure.in mysql-5.0.19/configure.in
--- mysql-5.0.19.orig/configure.in	2006-03-12 21:34:26.000000000 +0200
+++ mysql-5.0.19/configure.in	2006-03-12 21:35:17.000000000 +0200
@@ -2701,7 +2701,7 @@
 AC_SUBST([ndb_transporter_opt_objs])
 
 ndb_opt_subdirs=
-ndb_bin_am_ldflags="-static"
+ndb_bin_am_ldflags=""
 if test X"$have_ndb_test" = Xyes
 then
   ndb_opt_subdirs="test"
