diff -urNp -x '*.orig' mysql-5.0.96.org/patch_info/show_patches.info mysql-5.0.96/patch_info/show_patches.info
--- mysql-5.0.96.org/patch_info/show_patches.info	1970-01-01 01:00:00.000000000 +0100
+++ mysql-5.0.96/patch_info/show_patches.info	2022-10-18 08:28:16.729398106 +0200
@@ -0,0 +1,6 @@
+File=show_patches.patch
+Name=SHOW PATCHES
+Version=1.0
+Author=Jeremy Cole
+License=N/A
+Comment
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/Makefile.am mysql-5.0.96/sql/Makefile.am
--- mysql-5.0.96.org/sql/Makefile.am	2022-10-18 08:28:15.669398106 +0200
+++ mysql-5.0.96/sql/Makefile.am	2022-10-18 08:28:16.729398106 +0200
@@ -121,7 +121,7 @@ DEFS =			-DMYSQL_SERVER \
 			-DSHAREDIR="\"$(MYSQLSHAREdir)\"" \
 			@DEFS@
 
-BUILT_SOURCES =		sql_yacc.cc sql_yacc.h lex_hash.h
+BUILT_SOURCES =		sql_yacc.cc sql_yacc.h lex_hash.h patch_info.h
 EXTRA_DIST =		$(BUILT_SOURCES) nt_servc.cc nt_servc.h \
 			message.mc message.h message.rc MSG00001.bin \
 			examples/CMakeLists.txt CMakeLists.txt \
@@ -182,6 +182,8 @@ pkglib_LTLIBRARIES = sphinx/sphinx.la
 sphinx_sphinx_la_SOURCES = sphinx/snippets_udf.cc
 sphinx_sphinx_la_LDFLAGS = -module
 
+patch_info.h: patch_info.h.pl
+	$(PERL) $< > $@
 
 # Don't update the files from bitkeeper
 %::SCCS/s.%
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/Makefile.in mysql-5.0.96/sql/Makefile.in
--- mysql-5.0.96.org/sql/Makefile.in	2012-03-02 15:05:22.000000000 +0100
+++ mysql-5.0.96/sql/Makefile.in	2022-10-18 08:28:16.732731440 +0200
@@ -568,7 +568,7 @@ gen_lex_hash_SOURCES = gen_lex_hash.cc
 gen_lex_hash_LDADD = $(LDADD) $(CXXLDFLAGS)
 mysql_tzinfo_to_sql_SOURCES = mysql_tzinfo_to_sql.cc
 mysql_tzinfo_to_sql_LDADD = @MYSQLD_EXTRA_LDFLAGS@ $(LDADD) $(CXXLDFLAGS)
-BUILT_SOURCES = sql_yacc.cc sql_yacc.h lex_hash.h
+BUILT_SOURCES = sql_yacc.cc sql_yacc.h lex_hash.h patch_info.h
 EXTRA_DIST = $(BUILT_SOURCES) nt_servc.cc nt_servc.h \
 			message.mc message.h message.rc MSG00001.bin \
 			examples/CMakeLists.txt CMakeLists.txt \
@@ -1229,6 +1229,9 @@ lex_hash.h:	gen_lex_hash.cc lex.h
 		./gen_lex_hash$(EXEEXT) > $@-t
 		$(MV) $@-t $@
 
+patch_info.h: patch_info.h.pl
+	$(PERL) $< > $@
+
 # Don't update the files from bitkeeper
 %::SCCS/s.%
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/lex.h mysql-5.0.96/sql/lex.h
--- mysql-5.0.96.org/sql/lex.h	2012-03-02 15:04:12.000000000 +0100
+++ mysql-5.0.96/sql/lex.h	2022-10-18 08:28:16.732731440 +0200
@@ -367,6 +367,7 @@ static SYMBOL symbols[] = {
   { "PACK_KEYS",	SYM(PACK_KEYS_SYM)},
   { "PARTIAL",		SYM(PARTIAL)},
   { "PASSWORD",		SYM(PASSWORD)},
+  { "PATCHES",		SYM(PATCHES)},
   { "PHASE",            SYM(PHASE_SYM)},
   { "POINT",		SYM(POINT_SYM)},
   { "POLYGON",		SYM(POLYGON)},
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/mysql_priv.h mysql-5.0.96/sql/mysql_priv.h
--- mysql-5.0.96.org/sql/mysql_priv.h	2022-10-18 08:28:15.672731440 +0200
+++ mysql-5.0.96/sql/mysql_priv.h	2022-10-18 08:28:16.732731440 +0200
@@ -987,6 +987,7 @@ void mysqld_list_processes(THD *thd,cons
 int mysqld_show_status(THD *thd);
 int mysqld_show_variables(THD *thd,const char *wild);
 bool mysqld_show_storage_engines(THD *thd);
+bool mysqld_show_patches(THD *thd);
 bool mysqld_show_privileges(THD *thd);
 bool mysqld_show_column_types(THD *thd);
 bool mysqld_help (THD *thd, const char *text);
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/patch_info.h.pl mysql-5.0.96/sql/patch_info.h.pl
--- mysql-5.0.96.org/sql/patch_info.h.pl	1970-01-01 01:00:00.000000000 +0100
+++ mysql-5.0.96/sql/patch_info.h.pl	2022-10-18 08:28:16.732731440 +0200
@@ -0,0 +1,65 @@
+use strict;
+
+my $patch_info_path = '../patch_info';
+my $file = '';
+my $output = '';
+
+
+if (opendir(PATCH_DIR, $patch_info_path))
+{
+	while ((my $file = readdir(PATCH_DIR)))
+	{
+		open(PATCH_FILE, "<$patch_info_path/$file") || die("Unable to open $patch_info_path/$file ($!)");
+		my %fields;
+	
+		if ($file =~ /^\./)
+		{
+			next;
+		}	
+	
+		while (<PATCH_FILE>)
+		{
+			chomp;
+	
+			my ($key, $value) = split(/\s*=\s*/);
+			$fields{lc($key)} = $value;
+		}
+	
+		$output .= "{\"$fields{'file'}\", \"$fields{'name'}\", \"$fields{'version'}\", \"$fields{'author'}\", \"$fields{'license'}\",\"$fields{'comment'}\"},\n"
+	}
+}	
+
+print <<HEADER;
+
+/* Copyright (C) 2002-2006 MySQL AB
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
+
+#ifdef USE_PRAGMA_INTERFACE
+#pragma interface			/* gcc class implementation */
+#endif
+
+struct patch {
+	const char *file;
+	const char *name;
+	const char *version;
+	const char *author;
+	const char *license;
+	const char *comment;
+}patches[] = {
+$output
+{NULL, NULL, NULL, NULL}
+};	
+
+HEADER
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/sp_head.cc mysql-5.0.96/sql/sp_head.cc
--- mysql-5.0.96.org/sql/sp_head.cc	2012-03-02 15:04:07.000000000 +0100
+++ mysql-5.0.96/sql/sp_head.cc	2022-10-18 08:28:16.732731440 +0200
@@ -191,6 +191,7 @@ sp_get_flags_for_command(LEX *lex)
   case SQLCOM_SHOW_MUTEX_STATUS:
   case SQLCOM_SHOW_NEW_MASTER:
   case SQLCOM_SHOW_OPEN_TABLES:
+  case SQLCOM_SHOW_PATCHES:
   case SQLCOM_SHOW_PRIVILEGES:
   case SQLCOM_SHOW_PROCESSLIST:
   case SQLCOM_SHOW_SLAVE_HOSTS:
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/sql_lex.h mysql-5.0.96/sql/sql_lex.h
--- mysql-5.0.96.org/sql/sql_lex.h	2022-10-18 08:28:15.672731440 +0200
+++ mysql-5.0.96/sql/sql_lex.h	2022-10-18 08:28:16.732731440 +0200
@@ -96,6 +96,7 @@ enum enum_sql_command {
   SQLCOM_XA_COMMIT, SQLCOM_XA_ROLLBACK, SQLCOM_XA_RECOVER,
   SQLCOM_SHOW_PROC_CODE, SQLCOM_SHOW_FUNC_CODE,
   SQLCOM_SHOW_PROFILE, SQLCOM_SHOW_PROFILES,
+  SQLCOM_SHOW_PATCHES,
 
   /*
     When a command is added here, be sure it's also added in mysqld.cc
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/sql_parse.cc mysql-5.0.96/sql/sql_parse.cc
--- mysql-5.0.96.org/sql/sql_parse.cc	2022-10-18 08:28:15.676064773 +0200
+++ mysql-5.0.96/sql/sql_parse.cc	2022-10-18 08:28:16.732731440 +0200
@@ -4120,6 +4120,9 @@ end_with_restore_list:
   case SQLCOM_SHOW_STORAGE_ENGINES:
     res= mysqld_show_storage_engines(thd);
     break;
+  case SQLCOM_SHOW_PATCHES:
+    res= mysqld_show_patches(thd);
+    break;
   case SQLCOM_SHOW_PRIVILEGES:
     res= mysqld_show_privileges(thd);
     break;
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/sql_prepare.cc mysql-5.0.96/sql/sql_prepare.cc
--- mysql-5.0.96.org/sql/sql_prepare.cc	2012-03-02 15:04:08.000000000 +0100
+++ mysql-5.0.96/sql/sql_prepare.cc	2022-10-18 08:28:16.736064773 +0200
@@ -1818,6 +1818,7 @@ static bool check_prepared_statement(Pre
   case SQLCOM_SHOW_DATABASES:
   case SQLCOM_SHOW_PROCESSLIST:
   case SQLCOM_SHOW_STORAGE_ENGINES:
+  case SQLCOM_SHOW_PATCHES:
   case SQLCOM_SHOW_PRIVILEGES:
   case SQLCOM_SHOW_COLUMN_TYPES:
   case SQLCOM_SHOW_STATUS:
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/sql_show.cc mysql-5.0.96/sql/sql_show.cc
--- mysql-5.0.96.org/sql/sql_show.cc	2022-10-18 08:28:15.676064773 +0200
+++ mysql-5.0.96/sql/sql_show.cc	2022-10-18 08:28:16.736064773 +0200
@@ -22,6 +22,7 @@
 #include "sp.h"
 #include "sp_head.h"
 #include "sql_trigger.h"
+#include "patch_info.h"
 #include <my_dir.h>
 
 #ifdef HAVE_BERKELEY_DB
@@ -46,6 +47,47 @@ static int
 view_store_create_info(THD *thd, TABLE_LIST *table, String *buff);
 bool schema_table_store_record(THD *thd, TABLE *table);
 
+/***************************************************************************
+** List patches built into this release
+***************************************************************************/
+
+bool mysqld_show_patches(THD *thd)
+{
+  List<Item> field_list;
+	int i = 0;
+  Protocol *protocol= thd->protocol;
+  DBUG_ENTER("mysqld_show_patches");
+
+  field_list.push_back(new Item_empty_string("File", 255));
+  field_list.push_back(new Item_empty_string("Name", 50));
+  field_list.push_back(new Item_empty_string("Version", 10));
+  field_list.push_back(new Item_empty_string("Author", 50));
+  field_list.push_back(new Item_empty_string("License", 50));
+  field_list.push_back(new Item_empty_string("Comment", 32));
+
+  if (protocol->send_fields(&field_list, Protocol::SEND_NUM_ROWS | Protocol::SEND_EOF))
+    DBUG_RETURN(TRUE);
+
+	for (i = 0; patches[i].file; i++)
+	{
+	  protocol->prepare_for_resend();
+  	protocol->store(patches[i].file, system_charset_info);
+  	protocol->store(patches[i].name, system_charset_info);
+  	protocol->store(patches[i].version, system_charset_info);
+  	protocol->store(patches[i].author, system_charset_info);
+  	protocol->store(patches[i].license, system_charset_info);
+  	protocol->store(patches[i].comment, system_charset_info);
+
+  	if (protocol->write())
+    	DBUG_RETURN(TRUE);
+	}	
+
+   
+  send_eof(thd);
+  DBUG_RETURN(FALSE);
+
+}
+
 
 /***************************************************************************
 ** List all table types supported
diff -urNp -x '*.orig' mysql-5.0.96.org/sql/sql_yacc.yy mysql-5.0.96/sql/sql_yacc.yy
--- mysql-5.0.96.org/sql/sql_yacc.yy	2022-10-18 08:28:15.679398106 +0200
+++ mysql-5.0.96/sql/sql_yacc.yy	2022-10-18 08:28:16.736064773 +0200
@@ -859,6 +859,7 @@ bool my_yyoverflow(short **a, YYSTYPE **
 %token  PAGE_SYM
 %token  PARTIAL
 %token  PASSWORD
+%token  PATCHES
 %token  PARAM_MARKER
 %token  PHASE_SYM
 %token  POINTFROMTEXT
@@ -8061,7 +8062,7 @@ show:	SHOW
 	;
 
 show_param:
-         DATABASES wild_and_where
+        DATABASES wild_and_where
          {
            LEX *lex= Lex;
            lex->sql_command= SQLCOM_SELECT;
@@ -8162,6 +8163,10 @@ show_param:
 	    lex->sql_command= SQLCOM_SHOW_STORAGE_ENGINES;
 	    WARN_DEPRECATED("SHOW TABLE TYPES", "SHOW [STORAGE] ENGINES");
 	  }
+	| PATCHES
+	  {
+	    Lex->sql_command= SQLCOM_SHOW_PATCHES;
+	  }
 	| opt_storage ENGINES_SYM
 	  {
 	    LEX *lex=Lex;
@@ -9596,6 +9601,7 @@ keyword_sp:
 	| PAGE_SYM              {}
 	| PARTIAL		{}
 	| PASSWORD		{}
+	| PATCHES {}
         | PHASE_SYM             {}
 	| POINT_SYM		{}
 	| POLYGON		{}
