diff -urN mysql-4.0.12.org/client/mysqldump.c mysql-4.0.12/client/mysqldump.c
--- mysql-4.0.12.org/client/mysqldump.c	Wed Apr 16 00:26:02 2003
+++ mysql-4.0.12/client/mysqldump.c	Wed Apr 16 00:37:52 2003
@@ -1254,10 +1254,12 @@
   {
     if (opt_databases || opt_alldbs)
     {
+      char database_buff[NAME_LEN+3];
+      
       fprintf(md_result_file,"\n--\n-- Current Database: %s\n--\n", database);
       if (!opt_create_db)
 	fprintf(md_result_file,"\nCREATE DATABASE /*!32312 IF NOT EXISTS*/ %s;\n",
-		database);
+		quote_name(database, database_buff));
       fprintf(md_result_file,"\nUSE %s;\n", database);
     }
   }
