diff -ur mysql-5.0.41.org/configure.in mysql-5.0.41/configure.in
--- mysql-5.0.41.org/configure.in	2007-05-02 16:39:58.000000000 +0200
+++ mysql-5.0.41/configure.in	2007-05-10 15:02:12.075664162 +0200
@@ -779,19 +779,20 @@
 
 AC_CHECK_LIB(m, floor, [], AC_CHECK_LIB(m, __infinity))
 
-AC_CHECK_LIB(nsl_r, gethostbyname_r, [],
-  AC_CHECK_LIB(nsl, gethostbyname_r))
-AC_CHECK_FUNC(gethostbyname_r)
+AC_CHECK_FUNC(gethostbyname_r, [],
+	[AC_CHECK_LIB(nsl_r, gethostbyname_r, [],
+		[AC_CHECK_LIB(nsl, gethostbyname_r)])])
 
 AC_CHECK_FUNC(setsockopt, , AC_CHECK_LIB(socket, setsockopt))
-AC_CHECK_FUNC(yp_get_default_domain, ,
-  AC_CHECK_LIB(nsl, yp_get_default_domain))
 AC_CHECK_FUNC(p2open, , AC_CHECK_LIB(gen, p2open))
 # This may get things to compile even if bind-8 is installed
 AC_CHECK_FUNC(bind, , AC_CHECK_LIB(bind, bind))
 # Check if crypt() exists in libc or libcrypt, sets LIBS if needed
-AC_SEARCH_LIBS(crypt, crypt, AC_DEFINE(HAVE_CRYPT, 1, [crypt]))
-AC_SEARCH_LIBS(inet_aton, [socket nsl resolv])
+save_LIBS="$LIBS"
+AC_CHECK_LIB(crypt, crypt, [LIBS="-lcrypt $LIBS"; LIBCRYPT="-lcrypt"])
+AC_SUBST(LIBCRYPT)
+AC_CHECK_FUNC(crypt, AC_DEFINE([HAVE_CRYPT], [1], [crypt]))
+LIBS="$save_LIBS"
 
 # For the sched_yield() function on Solaris
 AC_CHECK_FUNC(sched_yield, , AC_CHECK_LIB(posix4, sched_yield))
diff -ur mysql-5.0.41.org/libmysql_r/Makefile.am mysql-5.0.41/libmysql_r/Makefile.am
--- mysql-5.0.41.org/libmysql_r/Makefile.am	2007-05-02 16:39:59.000000000 +0200
+++ mysql-5.0.41/libmysql_r/Makefile.am	2007-05-10 15:00:11.721844431 +0200
@@ -30,7 +30,7 @@
 include $(top_srcdir)/libmysql/Makefile.shared
 
 libmysqlclient_r_la_SOURCES = $(target_sources)
-libmysqlclient_r_la_LIBADD = $(target_libadd) $(yassl_las)
+libmysqlclient_r_la_LIBADD = $(target_libadd) $(yassl_las) -lpthread
 libmysqlclient_r_la_LDFLAGS = $(target_ldflags)
 
 # This is called from the toplevel makefile
diff -ur mysql-5.0.41.org/sql/Makefile.am mysql-5.0.41/sql/Makefile.am
--- mysql-5.0.41.org/sql/Makefile.am	2007-05-02 16:40:42.000000000 +0200
+++ mysql-5.0.41/sql/Makefile.am	2007-05-10 15:00:11.721844431 +0200
@@ -43,7 +43,7 @@
                         @innodb_system_libs@ \
 			@ndbcluster_libs@ @ndbcluster_system_libs@ \
 			$(LDADD)  $(CXXLDFLAGS) $(WRAPLIBS) @LIBDL@ \
-                        $(yassl_libs) $(openssl_libs)
+                        $(yassl_libs) $(openssl_libs) @LIBCRYPT@
 
 noinst_HEADERS =	item.h item_func.h item_sum.h item_cmpfunc.h \
 			item_strfunc.h item_timefunc.h item_uniq.h \
