diff -urNbB mysql-5.0.13-rc.org/sql/mysqld.cc mysql-5.0.13-rc/sql/mysqld.cc
--- mysql-5.0.13-rc.org/sql/mysqld.cc	2005-09-22 21:39:56.000000000 +0200
+++ mysql-5.0.13-rc/sql/mysqld.cc	2005-09-29 19:29:59.991454000 +0200
@@ -3183,7 +3183,7 @@
   init_ssl();
 
 #ifdef HAVE_LIBWRAP
-  libwrapName= my_progname+dirname_length(my_progname);
+  libwrapName= (char *)my_progname+dirname_length(my_progname);
   openlog(libwrapName, LOG_PID, LOG_AUTH);
 #endif
 
diff --git a/sql/sql_partition.cc b/sql/sql_partition.cc
index 201f64522a6..e23fe70ec86 100644
--- a/sql/sql_partition.cc
+++ b/sql/sql_partition.cc
@@ -284,7 +284,7 @@ bool partition_default_handling(TABLE *table, partition_info *part_info,
     }
   }
   part_info->set_up_defaults_for_partitioning(table->file,
-                                              (ulonglong)0, (uint)0);
+                                              NULL, 0U);
   DBUG_RETURN(FALSE);
 }
 

