--- mysql-5.7.36/cmake/build_configurations/compiler_options.cmake~	2021-09-07 07:26:47.000000000 +0200
+++ mysql-5.7.36/cmake/build_configurations/compiler_options.cmake	2021-11-29 07:46:49.013266334 +0100
@@ -73,7 +73,7 @@ IF(UNIX)
     EXECUTE_PROCESS(COMMAND ${CMAKE_CXX_COMPILER} -dumpversion
                     OUTPUT_VARIABLE GXX_VERSION)
     IF(GXX_VERSION VERSION_EQUAL 6.0 OR GXX_VERSION VERSION_GREATER 6.0)
-      STRING_PREPEND(COMMON_CXX_FLAGS "-std=gnu++03 ")
+      STRING_PREPEND(COMMON_CXX_FLAGS "-std=gnu++11 ")
     ENDIF()
     # Disable inline optimizations for valgrind testing to avoid false positives
     IF(WITH_VALGRIND)
